qiskit/pulse/instructions/call.py

```diff
@@ -0,0 +1,129 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Call instruction that represents calling a schedule as a subroutine."""
+
+from typing import Optional, Union, Dict, Tuple, Any, Set
+
+from qiskit.circuit.parameterexpression import ParameterExpression, ParameterValueType
+from qiskit.pulse.instructions import instruction
+
+# TODO This instruction should support ScheduleBlock when it's ready.
+
+
+class Call(instruction.Instruction):
+    """Pulse ``Call`` instruction.
+
+    This instruction represents a call to a referenced pulse subroutine (schedule).
+    Note that this instruction is not exposed to end users.
+    The ``Call`` instruction is created with a target subroutine and it behaves as a
+    standard pulse instruction within the ``Schedule``. No physical operation to
+    a quantum processor is triggered by this instruction. On the other hand,
+    this instruction provides a set of inline function (subroutine) and
+    parameters to the compiler, and the compiler can use these information to generate
+    better program representation, especially, when the inline function is used multiple times
+    within the main program.
+    """
+    # Prefix to use for auto naming.
+    prefix = 'subroutine-'
+
+    def __init__(self, subroutine, name: Optional[str] = None):
```

Comment:
I believe we should encourage parameters to be assigned at instantiation time for the `Call` instruction. This is a container for a schedule and a set of parameters that will later be bound to it. We should not encourage this class to be mutable. eg. something like,
```suggestion
    def __init__(self, subroutine, value_dict: Dict[param, any], name: Optional[str] = None):
```

Reply:
I added value dict to the constructor, but still this instruction is mutable. We may want to define a schedule template by calling another subroutine which is parameterized, because parameters may drift from time to time. 

For example, writing this for every subroutine call is bit annoying.
```python
pulse.call(subroutine, amp=subroutine.get_parameters('amp')[0], sigma=subroutine.get_parameters('sigma')[0], ...)
```
So I'd like to keep an option that enable us to insert subroutine without parameter specification.